x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityMigrationExecutor.java

```diff
@@ -0,0 +1,114 @@
+import static org.elasticsearch.xpack.security.support.SecurityMigrations.isEligibleSecurityMigration;
+
+public class SecurityMigrationExecutor extends PersistentTasksExecutor<SecurityMigrationTaskParams> {
+
+    private static final Logger logger = LogManager.getLogger(SecurityMigrationExecutor.class);
+    private final SecurityIndexManager securityIndexManager;
+    private final Client client;
+    private final FeatureService featureService;
+    private final ClusterService clusterService;
+    private final TreeMap<Integer, SecurityMigrations.SecurityMigration> migrationByVersion;
+
+    public SecurityMigrationExecutor(
+        String taskName,
+        Executor executor,
+        SecurityIndexManager securityIndexManager,
+        Client client,
+        FeatureService featureService,
+        ClusterService clusterService,
+        TreeMap<Integer, SecurityMigrations.SecurityMigration> migrationByVersion
+    ) {
+        super(taskName, executor);
+        this.securityIndexManager = securityIndexManager;
+        this.client = client;
+        this.featureService = featureService;
+        this.clusterService = clusterService;
+        this.migrationByVersion = migrationByVersion;
+    }
+
+    @Override
+    protected void nodeOperation(AllocatedPersistentTask task, SecurityMigrationTaskParams params, PersistentTaskState state) {
+        applyOutstandingMigrations(task, params.getMigrationVersion(), ActionListener.wrap((res) -> task.markAsCompleted(), (exception) -> {
+            logger.warn("Security migration failed: " + exception);
+            task.markAsFailed(exception);
+        }));
+    }
+
+    private void applyOutstandingMigrations(AllocatedPersistentTask task, int currentMigrationVersion, ActionListener<Void> listener) {
+        if (task.isCancelled()) {
+            listener.onFailure(new TaskCancelledException("Security migration task cancelled"));
+            return;
+        }
+        if (migrationByVersion.lastKey() > currentMigrationVersion) {
+            Map.Entry<Integer, SecurityMigrations.SecurityMigration> migrationEntry = migrationByVersion.higherEntry(
+                currentMigrationVersion
+            );
+            int clusterMappingsVersion = loadIndexMappingVersion(securityIndexManager.aliasName(), clusterService.state());
+            SecurityMigrations.SecurityMigration migration = migrationEntry.getValue();
+
+            // Check if all nodes can support feature and that the cluster is on a compatible mapping version
+            if (isEligibleSecurityMigration(clusterMappingsVersion, featureService, clusterService.state(), migration)) {
+                migration.migrate(
+                    securityIndexManager,
+                    client,
+                    ActionListener.wrap(
+                        response -> updateMigrationVersion(
+                            migrationEntry.getKey(),
+                            securityIndexManager.getConcreteIndexName(),
```

Comment:
👍 I think this is optimal, disregard my verbal mention of using the `IndexNameExpressionResolver`.

Reply:
Ok cool! Thanks for looking into it!